File: c/algorithms/majority.cpp
# include <stdio.h>
int candidate(int *a,int n,int m){
	int i = m ;
    int c = a[m];
    int count = 1;
    while(i < n && count > 0){
    	if(a[++i] == c)	count ++;
    	else	count--;
    }
    if(i == n) return c ;
    else return candidate(a,n,i+1);
}
int majority(int *a,int n){
	int c = candidate(a,n-1,0);
	int count = 0;
	for(int i = 0; i<n ; i++){
		if (c == a[i]){
		count++;
		}
	}
	if(count > n/2) return c;
	else return NULL;
}
int main(){
	int a[8] ={2,4,1,4,4,4,6,4};
	int m = majority(a,8);
	printf("多数元素是：%d\n",m);
	return 0;
} 

File: JavaExercise/Java_stream/src/Majority.java

public class Majority {
	public static void main(String[] args) {
		int A[] = {4,9,9,9,9,6};
		int B[] = {9,4,9,7,9,6};
		majority(A);
		majority(B);

		
	}
	public static void majority(int A[]){
		int n = A.length-1;
		int c = candidate(A,0,n);
		int count = 0;
		for(int j =0 ; j<n ;j++){
			if(A[j] == c){
				count++;
			}
		}
		if(count > (n+1)/2){
			System.out.println("多数为；"+c);
		}
		else{
			System.out.println("没有多数");
		}
	
	}
	
	public static int candidate(int A[],int m,int n){
		int j= m;
		int c= A[m];
		int count = 1;
		while (j<n && count >0){
			j++;
			if (A[j] == c ){
				count++;
			}
			else{
				count--;
			}
		}
		if(j == n){
			return c;
		}
		else{
			return candidate(A,j+1,n);
		}
	}
	
}